src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+    virtual void abort() = 0;
+
+    //create a new instance of this CallData object, with the same type
+    //(same template parameters) as original. This is called when a CallData
+    //object starts processing a request. Creating a new instance allows the 
+    //server to handle additional requests while the first is being processed
+    virtual std::shared_ptr<Processor> clone() = 0;
+
+    //true if this object has finished processing the request. Object will be
+    //deleted once this function returns true
+    virtual bool isFinished() = 0;
+};
+
+namespace {
+
+//helper function. strips port from endpoint string
+std::string getEndpoint(std::string const& peer)
+{
+    std::size_t first = peer.find_first_of(":");
+    std::size_t last = peer.find_last_of(":");
+    std::string peer_clean(peer);
+    if(first != last)
+    {
+        peer_clean = peer.substr(first+1);
+    }
+    return peer_clean;
+}
+
+} //namespace
+
+class GRPCServerImpl final {
+
+    private:
+
+    //list of current RPC requests being processed or listened for
+    std::list<std::shared_ptr<Processor>> requests_;
+
+    //CompletionQueue returns events that have occurred, or events that have
+    //been cancelled
+    std::unique_ptr<grpc::ServerCompletionQueue> cq_;
+
+    //The gRPC service defined by the .proto files
+    rpc::v1::XRPLedgerAPIService::AsyncService service_;
+
+    //gRPC server
+    std::unique_ptr<grpc::Server> server_;
+
+    //referernce to ripple::Application
+    Application& app_;
+
+    //address of where to run the server
+    //this is the default value, if ip and port are not in config
+    std::string server_address_ = "0.0.0.0:50051";
+
+
+    //typedef for function to bind a listener
+    template <class Request, class Response>
+    using BindListener = std::function<void(
+            rpc::v1::XRPLedgerAPIService::AsyncService&,
+            grpc::ServerContext*,
+            Request*,
+            grpc::ServerAsyncResponseWriter<Response>*,
+            grpc::CompletionQueue*,
+            grpc::ServerCompletionQueue*,
+            void*)>;
+
+    //typedef for actual handler (that populates a response)
+    template <class Request, class Response>
+    using Handler = std::function<std::pair<Response,grpc::Status>(
+            RPC::ContextGeneric<Request>&)>;
+
+    public:
+    ~GRPCServerImpl() {}
+
+    GRPCServerImpl(Application& app);
+  
+    void shutdown()
+    {
+        server_->Shutdown();
+        // Always shutdown the completion queue after the server.
+        cq_->Shutdown(); 
+    }
+  
+    //setup the server and listeners
+    void start();
+
+    //the main event loop
+    void handleRpcs();
+  
+    private:
+  
+    // Class encompasing the state and logic needed to serve a request.
+    template <class Request,class Response>
+    class CallData :
+        public Processor, 
+        public std::enable_shared_from_this<CallData<Request,Response>>
```

Comment:
I don't think we need to inherit `enabled_shared_from_this`. I'd change `process` so it's a static function and pass in the pointer as a parameter.

Reply:
@cjcobb23 I'm resolving this - I'm fine keeping `enabled_shared_from_this`.